metricbeat/mb/lightmetricset.go

```diff
@@ -0,0 +1,118 @@
+	Module  string
+	Default bool `config:"default"`
+	Input   struct {
+		Module    string      `config:"module" validate:"required"`
+		MetricSet string      `config:"metricset" validate:"required"`
+		Defaults  interface{} `config:"defaults"`
+	} `config:"input" validate:"required"`
+}
+
+// Registration obtains a metric set registration for this light metric set, this registration
+// contains a metric set factory that reprocess metric set creation taking into account the
+// light metric set defaults
+func (m *LightMetricSet) Registration(r *Register) (MetricSetRegistration, error) {
+	registration, err := r.metricSetRegistration(m.Input.Module, m.Input.MetricSet)
+	if err != nil {
+		return registration, errors.Wrapf(err,
+			"failed to start light metricset '%s/%s' using '%s/%s' metricset as input",
+			m.Module, m.Name,
+			m.Input.Module, m.Input.MetricSet)
+	}
+
+	originalFactory := registration.Factory
+	registration.IsDefault = m.Default
+
+	// Light modules factory has to override defaults and reproduce builder
+	// functionality with the resulting configuration, it does:
+	// - Override defaults
+	// - Call module factory if registered (it wouldn't have been called
+	//   if light module is really a registered mixed module)
+	// - Call host parser if defined (it would have already been called
+	//   without the light module defaults)
+	// - Finally, call the original factory for the registered metricset
+	registration.Factory = func(base BaseMetricSet) (MetricSet, error) {
+		// Override default config on base module and metricset
+		base.name = m.Name
+		baseModule, err := m.baseModule(base.module)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create base module for light module")
+		}
+		base.module = baseModule
+
+		// Run module factory if registered, it will be called once per
+		// metricset, but it should be idempotent
+		moduleFactory := r.moduleFactory(m.Input.Module)
+		if moduleFactory != nil {
+			module, err := moduleFactory(*baseModule)
+			if err != nil {
+				return nil, errors.Wrapf(err, "module factory for module '%s' failed while creating light metricset '%s/%s'", m.Input.Module, m.Module, m.Name)
+			}
+			base.module = module
+		}
+
+		// At this point host parser was already run, we need to run this again
+		// with the overriden defaults
+		if registration.HostParser != nil {
+			base.hostData, err = registration.HostParser(base.module, base.host)
+			if err != nil {
+				return nil, errors.Wrapf(err, "host parser failed on light metricset factory for '%s/%s'", m.Module, m.Name)
+			}
+			base.host = base.hostData.Host
+		}
+
+		return originalFactory(base)
+	}
+
+	return registration, nil
+}
+
+// baseModule does the configuration overrides in the base module configuration
+// taking into account the light metric set default configurations
+func (m *LightMetricSet) baseModule(from Module) (*BaseModule, error) {
+	baseModule := BaseModule{
+		name:   m.Module,
+		config: from.Config(),
```

Comment:
It worked as expected on my tests, but yes, it doesn't look safe, I will try to get a better copy of this module config.

> We call 'UnpackConfig`with target`config` a few lines down from here. Could this call overwrite entries in Hosts, MetricSets and Query?
> 
> What is the expected result after UnpackConfig if Hosts/MetricSets/Query was not empty?

Hosts and MetricSets shouldn't be changed here, query could be, in this case the map should be merged. It may require more tests to cover these cases.

Reply:
This was not safe, and this was actually not needed as we copy from the raw config later. I have extended the tests to check that original base module is not modified, and that we apply the overrides in the case of query.